zenith_utils/src/http/endpoint.rs

```diff
@@ -0,0 +1,164 @@
+    Ok(res)
+}
+
+async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body>, ApiError> {
+    SERVE_METRICS_COUNT.inc();
+
+    let mut buffer = vec![];
+    let encoder = TextEncoder::new();
+    let metrics = zenith_metrics::gather();
+    encoder.encode(&metrics, &mut buffer).unwrap();
+
+    let response = Response::builder()
+        .status(200)
+        .header(CONTENT_TYPE, encoder.format_type())
+        .body(Body::from(buffer))
+        .unwrap();
+
+    Ok(response)
+}
+
+pub fn make_router() -> RouterBuilder<hyper::Body, ApiError> {
+    Router::builder()
+        .middleware(Middleware::post_with_info(logger))
+        .get("/metrics", prometheus_metrics_handler)
+        .err_handler(error::handler)
+}
+
+pub fn attach_openapi_ui(
+    router_builder: RouterBuilder<hyper::Body, ApiError>,
+    spec: &'static [u8],
+    spec_mount_path: &'static str,
+    ui_mount_path: &'static str,
+) -> RouterBuilder<hyper::Body, ApiError> {
+    router_builder.get(spec_mount_path, move |_| async move {
+        Ok(Response::builder().body(Body::from(spec)).unwrap())
+    }).get(ui_mount_path, move |_| async move {
+        Ok(Response::builder().body(Body::from(format!(r#"
+            <!DOCTYPE html>
+            <html lang="en">
+            <head>
+            <title>rweb</title>
+            <link href="https://cdn.jsdelivr.net/npm/swagger-ui-dist@3/swagger-ui.css" rel="stylesheet">
+            </head>
+            <body>
+                <div id="swagger-ui"></div>
+                <script src="https://cdn.jsdelivr.net/npm/swagger-ui-dist@3/swagger-ui-bundle.js" charset="UTF-8"> </script>
+                <script>
+                    window.onload = function() {{
+                    const ui = SwaggerUIBundle({{
+                        "dom_id": "\#swagger-ui",
+                        presets: [
+                        SwaggerUIBundle.presets.apis,
+                        SwaggerUIBundle.SwaggerUIStandalonePreset
+                        ],
+                        layout: "BaseLayout",
+                        deepLinking: true,
+                        showExtensions: true,
+                        showCommonExtensions: true,
+                        url: "{}",
+                    }})
+                    window.ui = ui;
+                }};
+            </script>
+            </body>
+            </html>
+        "#, spec_mount_path))).unwrap())
+    })
+}
+
+fn parse_token(header_value: &str) -> Result<&str, ApiError> {
+    // header must be in form Bearer <token>
+    let (prefix, token) = header_value.split_once(' ').ok_or(ApiError::Unauthorized(
+        "malformed authorization header".to_string(),
+    ))?;
+    if prefix != "Bearer" {
+        Err(ApiError::Unauthorized(
+            "malformed authorization header".to_string(),
+        ))?
+    }
+    Ok(token)
+}
+
+pub fn auth_middleware<B: hyper::body::HttpBody + Send + Sync + 'static>(
```

Comment:
What does this do? What is the argument?

Reply:
Middleware is a layer between our custom logic (i.e. get/post/etc handlers) and the depths of a web framework. In this case, it is responsible for authentication, and the argument you're referencing to is used to extract auth data from a request data type in an app-state-agnostic way. Each `Request<Body>` keeps a reference to the app's state.